### DISCUSSION TITLE
Code duplication in refinements

### DISCUSSION THREAD

**Discussion by eventualbuddha** [colinhacks/zod#346] [Ideas]
I've been using zod for parsing JSON objects in my project, and I'm trying to replace some custom validation with a zod refinement. However, I've found I'm unable to provide the error messages I want in zod v1. I noticed that v3 added the ability to customize the error message by passing a function instead of a static message. That's good! However, in my case having the "check" and "message" be two different things would mean duplicating the validation. [Here's what I'm doing now](https://github.com/votingworks/vxsuite/blob/3c65ee80a62382370e4c829c9db2fcfedfbd18de/libs/types/src/election.ts#L554-L659), for reference.

Here's a simple example below. Note that I have to call `findDuplicateIds` multiple times:

```ts
const Person = z.object({
  id: z.string(),
})

const People = z.array(Person).refine(
  (people) => !findDuplicateIds(people).next().done,
  (people) => ({
    message: `person IDs must be unique, but found duplicates: ${[...findDuplicateIds(people)].join(', ')}`
  })
)

function* findDuplicateIds<T extends { id: unknown }>(
  identifiables: Iterable<T>
): Generator<T['id']> {
  const knownIds = new Set<T['id']>()

  for (const { id } of identifiables) {
    if (knownIds.has(id)) {
      yield id
    } else {
      knownIds.add(id)
    }
  }
}
```

Ideally I'd be able to combine the two, returning errors if there are any and returning nothing if it's valid:

```ts
const People = z.array(Person).refine(
  (people) => {
    const duplicateIds = [...findDuplicateIds(people)]
    if (duplicateIds.length) {
      return {
        message: `person IDs must be unique, but found duplicates: ${duplicateIds.join(', ')}`
      }
    }
  }
)
```

**Comment by colinhacks** [comment]
You can do this! It's called `superRefine` and it's not yet documented (but its what `.refine()` uses under the hood). Here's how to do what you're looking for:

```ts
const People = z.array(Person).superRefine((people, ctx) => {
    const duplicateIds = [...findDuplicateIds(people)];
    if (duplicateIds.length) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        message: `person IDs must be unique, but found duplicates: ${duplicateIds.join(
          ", "
        )}`,
      });
    }
  });
```

Instead of returning `{ message: string }` you just call `ctx.addIssue` multiple times (which lets you create multiple issues in a single `superRefine` call). Haven't documented this yet because it's a bit complex and it's only useful to power-users. So congrats on being a power user!

**Reply by eventualbuddha** [reply]
Great, thanks!
